### PR TITLE
Do error cleanups

### PR DESCRIPTION
**Pull request by tcharding**
Do error cleanups, and add a script to help find missing `non_exhaustive` on error types

- Patch 1: Audit the codebase and put `non_exhaustive` on all  error types. 
- Patch 2: Implement `std::error::Error::source` for `bip152::Error`
- Patch 3: Implement `Display` and `std::error::Error::source` for `MerkleBlockError`
- Patch 4: Move rustdocs to above attributes on one error type
- ~Patch 5: Add a python script to `contrib` that checks the codebase for missing non exhaustive on the line above the results of regex `pub enum .*Error`~

I removed the Python script patch, I can't be bothered working on this for now but the clean ups and `non_exhaustive` additions are useful IMO.

### labels
- I added 'API break', I think the `non_exhaustive` addition requires major bump but not sure?
- release notes mention is just for the `non_exhaustive` patch.

### CODE REVIEW

**Comment by tcharding** [comment]
The python is pretty dirty, I can probably clean it up if folks like this. It took me long enough to write as is so will wait for review before spending more time on it.

**Review by Kixunil — COMMENTED** [review]
ACK 65abfaf7a60b75a945c24de2d3e2f7ccc8e84d46

(Yes, I meant everything but python script, yes I meant to not choose "approve" in GH)

I'd like the Python script to be a bit more robust if you're going to add it.

File: contrib/non-exhaustive.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        file = open(fname)
+
+        last_line = ""
+        missing = []
+
+        while True:
+                next_line = file.readline()
+
+                if re.search("pub enum .*Error", next_line):
+                        if not re.search("non_exhaustive", last_line):
```
[the remark this answers is not in the document]
Comment (tcharding) [reply contrib/non-exhaustive.py:25]: lolz, I love it. I admit my python is pretty rudimentary, one advantage it has is I was imagining we would add whitelisting to the script if/when we implement error types that do not require `non_exhaustive`. But realy, its a dirty tool to help us, so I'm cool to use either. Mad props for the one-liner, I did try to do it in bash before I wrote the python :)

File: src/util/bip152.rs [hunk 2]
```diff
@@ -38,7 +38,16 @@ impl fmt::Display for Error {
...
 
 #[cfg(feature = "std")]
-impl error::Error for Error {}
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use self::Error::*;
+
+        match *self {
+            UnknownVersion | InvalidPrefill => None,
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/util/bip152.rs:47]: oh, my bad.

File: contrib/non-exhaustive.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply contrib/non-exhaustive.py:25]: Oh, forgot to mention that some error types don't end in `Error` We should search all `fn ident({not regex}) -> Result<{not regex}, E>`

The language is not regular, so we can't use regular expressions. Maybe we should use a Rust program and the `syn` crate instead?

[the remark this answers is not in the document]
Comment (Kixunil) [reply contrib/non-exhaustive.py:25]: Damn, shouldn't have come up with the idea, now I'm tempted into having a general analysis tool to enforce crate policy...

**Comment by tcharding** [comment]
> (Yes, I meant everything but python script, yes I meant to not choose "approve" in GH)

> I'd like the Python script to be a bit more robust if you're going to add it.

Please re-review when you get time @Kixunil, script has been removed. No other changes IIRC.

File: src/util/merkleblock.rs [hunk 3]
```diff
@@ -53,20 +55,45 @@ use crate::consensus::encode::{self, Decodable, Encodable};
 use crate::util::merkleblock::MerkleBlockError::*;
 use crate::{Block, BlockHeader};
 
-/// An error when verifying the merkle block
+/// An error when verifying the merkle block.
 #[derive(Clone, PartialEq, Eq, Debug)]
 #[non_exhaustive]
 pub enum MerkleBlockError {
-    /// When header merkle root don't match to the root calculated from the partial merkle tree
+    /// Merkle header root doesn't match to the root calculated from the partial merkle tree.
```
Comment (Kixunil) [inline src/util/merkleblock.rs:62]: Shouldn't this be "Merkle root in the header doesn't match..."? Never heard of "merkle header root".

**Review by Kixunil — COMMENTED** [review]
I don't feel like creating noise in git history with followup commits, so I'd rather re-review, if you're OK with the suggested change.

File: src/util/merkleblock.rs (hunk 3 above)
Comment (tcharding) [reply src/util/merkleblock.rs:62]: Sure thing, I have no idea. Will use as suggested.

**Comment by tcharding** [comment]
Changes in force push: just the doc improvement suggested above.

**Review by Kixunil — APPROVED** [review]
ACK abc014a079c51b8562a8a58927193e8c8f45aa36

**Review by apoelstra — APPROVED** [review]
ACK abc014a079c51b8562a8a58927193e8c8f45aa36

**Pull request merged**